pytorch_lightning/utilities/auto_restart.py

```diff
@@ -0,0 +1,236 @@
+        return self.sampler.batch_indices
+
+    def __iter__(self) -> Iterator[List[int]]:
+        # setup local counter
+        restart_counter = 0
+
+        # iteration over wrapped sampler
+        for batch in self.sampler:
+
+            # if we are restarting, we will fastforward the batches
+            if self.restarting:
+
+                # the state dict is cached until workers are made available through the DataLoader
+                if self._cached_state_dict is not None and self.worker_id in self._cached_state_dict:
+
+                    # reload the current state dict
+                    self.load_state_dict(self._cached_state_dict, workers_initialized=True)
+                    self._cached_state_dict = None
+
+                # increment counter
+                restart_counter += 1
+
+                # if restart counter matches the current iteration, we should stop restarting
+                if restart_counter == self._current_iteration:
+                    self.restarting = False
+            else:
+                self._current_iteration += 1
+
+                # yield the batch
+                yield batch
+
+        self.reset_iteration()
+
+    def reset_iteration(self) -> None:
+        self._current_iteration = 0
+
+    def __len__(self) -> int:
+        return len(self.sampler)
+
+    def _compute_current_iteration(self, num_batches_processed: Optional[int] = None) -> int:
+        """
+        This function is used to compute the effective iteration.
+        As DataLoader can perform ``prefecthing`` or training can fail while processing a batch,
+        the current iteration needs to be computed using the ``num_batches_processed`` processed information.
+        """
+        if num_batches_processed is not None:
+            current_iteration = num_batches_processed
+        else:
+            current_iteration = self._current_iteration
+
+        if self._dataloader_batch_size:
+            current_iteration *= self._dataloader_batch_size
+
+        return current_iteration
+
+    def state_dict(self, num_batches_processed: Optional[int] = None) -> Dict[int, Dict[str, int]]:
+        """ Returns the state of the sampler in the current worker. The worker id indexes the state dict."""
+        return {self.worker_id: {"current_iteration": self._compute_current_iteration(num_batches_processed)}}
+
+    def load_state_dict(self, state_dict: Dict[int, Any], workers_initialized: bool = False) -> None:
+        """
+        Loads the saved state for the wrapped sampler.
+        If the ``state_dict`` contains multiple states, it means there were multiple workers.
+        The state will be cached and fully reloaded (fast-forward) the first time :meth:`__iter__` is called.
+        """
+        # as workers aren't available, the ``state_dict``` is cached until workers are made available.
+        if len(state_dict) > 1 and not workers_initialized:
+            self._cached_state_dict = state_dict
```

Comment:
should we deepcopy this, since we are caching what gets passed in?

Reply:
Yes, good idea !

Reply:
Done !